### PR TITLE
create reader creator decorators: shuffle, compose, chain

### PR DESCRIPTION
**Pull request by helinwang**
Fixes: https://github.com/PaddlePaddle/Paddle/issues/1380

### CODE REVIEW

File: python/paddle/reader/decorator.py [hunk 1]
```diff
@@ -12,25 +12,142 @@
...
+    Args:
+        reader_creator: the original reader creator whose output will be
+            shuffled.
+        buf_size: shuffle buffer size.
+
+    Returns:
+        the new reader creator whose output is shuffled.
+    """
+
+    def create_reader_creator():
```
Comment (wangkuiyi) [inline python/paddle/reader/decorator.py]: create_reader_creator ==> reader

File: python/paddle/reader/decorator.py [hunk 2]
```diff
@@ -12,25 +12,142 @@
...
+    [0, 0, 0, 1, 1, 1, 2, 2, 2]
+
+    Args:
+        readers_creators: input reader creators
+
+    Returns:
+        the new data reader creator.
+    """
+
+    def create_reader_creator():
```
Comment (wangkuiyi) [inline python/paddle/reader/decorator.py]: create_reader_creator ==> reader

File: python/paddle/reader/decorator.py [hunk 3]
```diff
@@ -12,25 +12,142 @@
...
+    """
+    check_alignment = kwargs.pop('check_alignment', True)
+
+    def make_tuple(x):
+        if isinstance(x, tuple):
+            return x
+        else:
+            return (x, )
+
+    def create_reader_creator():
```
Comment (wangkuiyi) [inline python/paddle/reader/decorator.py]: create_reader_creator ==> reader

File: python/paddle/reader/decorator.py [hunk 4]
```diff
@@ -12,25 +12,142 @@
...
+            will be discarded. Defaults to True.
+
+    Returns:
+        the new data reader creator.
+
+    Raises:
+        ComposeNotAligned: outputs of reader creators are not aligned.
+            Will not raise when check_alignment is set to False.
+    """
+    check_alignment = kwargs.pop('check_alignment', True)
```
Comment (wangkuiyi) [inline python/paddle/reader/decorator.py]: 这里是不是不需要允许用户定制了，而是改为必须check alignment？

File: python/paddle/reader/decorator.py [hunk 5]
```diff
@@ -43,8 +160,8 @@ def read_worker(r, q):
             q.put(d)
         q.put(end)
 
-    def create_reader():
-        r = reader()
+    def create_reader_creator():
```
Comment (wangkuiyi) [inline python/paddle/reader/decorator.py]: create_reader_creator ==> reader

File: python/paddle/reader/tests/decorator_test.py [hunk 6]
```diff
@@ -17,15 +17,18 @@
 
 
 def reader_10(dur):
-    for i in range(10):
-        time.sleep(dur)
-        yield i
+    def reader():
```
Comment (wangkuiyi) [inline python/paddle/reader/tests/decorator_test.py]: reader_10 ==> reader_creator_10

File: python/paddle/reader/tests/decorator_test.py [hunk 7]
```diff
@@ -34,17 +37,73 @@ def test_read(self):
...
+                self.assertLess(elapsed_time, 0.05)
             last_time = time.time()
 
 
+class TestCompose(unittest.TestCase):
+    def test_compse(self):
+        a = reader_10(0)
+        b = reader_10(0)
+        c = paddle.reader.compose(a, b)
+        for idx, e in enumerate(c()):
```
Comment (wangkuiyi) [inline python/paddle/reader/tests/decorator_test.py]: 这里如果写成

```python
for idx, ele in enumerate(
        paddle.reader.compose(
            reader_10(0),
            reader_10(0))):
    self.assertEqual(ele, (idx, idx))
```

是不是更能展示为什么需要定义一个decorator的参数是readers——因为可以让creator调用作为decorator的参数？

File: python/paddle/reader/tests/decorator_test.py [hunk 8]
```diff
@@ -17,15 +17,18 @@
 
 
 def reader_10(dur):
-    for i in range(10):
-        time.sleep(dur)
-        yield i
+    def reader():
+        for i in range(10):
+            time.sleep(dur)
```
Comment (wangkuiyi) [inline python/paddle/reader/tests/decorator_test.py]: I think we should have a comment for the time.sleep call:

```python
# This invocation helps testing paddle.reader.buffer
```

File: python/paddle/reader/tests/decorator_test.py [hunk 9]
```diff
@@ -34,17 +37,73 @@ def test_read(self):
...
+                total += 1
+        # expecting 10, not 20
+        self.assertEqual(total, 10)
+
+    def test_compose_not_aligned_no_check(self):
+        a = reader_10(0)
+        b = paddle.reader.chain(a, a)
+        c = paddle.reader.compose(a, b, check_alignment=False)
+        total = 0
+        for e in c():
```
Comment (wangkuiyi) [inline python/paddle/reader/tests/decorator_test.py]:
```python
for ele in paddle.reader.compose(
        paddle.reader.chain(reader_creator_10(0), reader_creator_10(0)),
        reader_creator_10(0)):
    total += 1
```

这样写能更让compose的两个参数的查边显得更清楚一点吗？

File: python/paddle/reader/tests/decorator_test.py [hunk 10]
```diff
@@ -34,17 +37,73 @@ def test_read(self):
...
+        self.assertEqual(total, 10)
+
+
+class TestChain(unittest.TestCase):
+    def test_chain(self):
+        a = reader_10(0)
+        b = reader_10(0)
+        c = paddle.reader.chain(a, b)
+        idx = 0
+        for e in c():
```
Comment (wangkuiyi) [inline python/paddle/reader/tests/decorator_test.py]:
```python
for ele in paddle.reader.chain(
        reader_creator_10(0),
        reader_creator_10(0)):
    self.assertEqual(ele, idx % 10)
    idx += 1
```

这样写可以展示给用户如何利用creator和decorator来减少代码行数吗？

File: python/paddle/reader/decorator.py (hunk 4 above)
Comment (helinwang) [inline python/paddle/reader/decorator.py]: check alignment在compose image和label的时候是必须的，但是有些情况不需要：
比如说这里compose了true image和无限个randome image. https://github.com/PaddlePaddle/Paddle/pull/1412/files#diff-053c4dca85f4c88cc67b76fc4ed1344aR94

File: python/paddle/reader/decorator.py (hunk 1 above)
Comment (helinwang) [inline python/paddle/reader/decorator.py]: Done.

File: python/paddle/reader/decorator.py (hunk 2 above)
Comment (helinwang) [inline python/paddle/reader/decorator.py]: Done.

File: python/paddle/reader/decorator.py (hunk 3 above)
Comment (helinwang) [inline python/paddle/reader/decorator.py]: Done.

File: python/paddle/reader/decorator.py (hunk 5 above)
Comment (helinwang) [inline python/paddle/reader/decorator.py]: Done.

File: python/paddle/reader/tests/decorator_test.py (hunk 6 above)
Comment (helinwang) [inline python/paddle/reader/tests/decorator_test.py]: Done.

File: python/paddle/reader/tests/decorator_test.py (hunk 8 above)
Comment (helinwang) [inline python/paddle/reader/tests/decorator_test.py]: Done.

File: python/paddle/reader/tests/decorator_test.py (hunk 7 above)
Comment (helinwang) [inline python/paddle/reader/tests/decorator_test.py]: 有道理！需要改成
```python
for idx, ele in enumerate(
        paddle.reader.compose(
            reader_10(0),
            reader_10(0)))():
    self.assertEqual(ele, (idx, idx))
```
才能编译通过（多一个`()`调用reader)，所以改成了
```python
        reader = paddle.reader.compose(reader_creator_10(0), reader_creator_10(0))
        for idx, e in enumerate(reader()):
            self.assertEqual(e, (idx, idx))
```

File: python/paddle/reader/tests/decorator_test.py (hunk 9 above)
Comment (helinwang) [inline python/paddle/reader/tests/decorator_test.py]: 👍，done.

File: python/paddle/reader/tests/decorator_test.py (hunk 10 above)
Comment (helinwang) [inline python/paddle/reader/tests/decorator_test.py]: Done.

File: python/paddle/reader/decorator.py [hunk 11]
```diff
@@ -12,18 +12,134 @@
...
+
+    If input readers output following data entries:
+    [0, 0, 0]
+    [1, 1, 1]
+    [2, 2, 2]
+    The chained reader will output:
+    [0, 0, 0, 1, 1, 1, 2, 2, 2]
+
+    Args:
+        readerss: input readers.
```
Comment (reyoung) [inline python/paddle/reader/decorator.py]: typo

readerss => readers

File: python/paddle/reader/decorator.py [hunk 12]
```diff
@@ -12,18 +12,134 @@
...
+        for r in readers:
+            rs.append(r())
+
+        for e in itertools.chain(*rs):
+            yield e
+
+    return reader
+
+
+class ComposeNotAligned:
```
Comment (reyoung) [inline python/paddle/reader/decorator.py]:
```
class ComposeNotAligned(ValueError):
    pass
```

There are several base error types in Python. [ValueError](https://docs.python.org/2/library/exceptions.html#exceptions.ValueError) means the argument is an inappropriate value but right type.

Use these base error types can help users figure out what's wrong in their code.

And more pythonic way is:

```python
raise ValueError("The reader in compose is not aligned").
```

Comment (helinwang) [inline python/paddle/reader/decorator.py]: @reyoung in
```python
raise ValueError("The reader in compose is not aligned").
```
how can user know that this exception is align error? E.g., before user can do
```python
if isinstance(e, ComposeNotAligned):
  # handle not aligned error
```
Or, it is in python exceptions typically not handled according to what exception it is?

Comment (reyoung) [inline python/paddle/reader/decorator.py]: Or at least, Make ComposeNotAligned as a subclass of ValueError.

File: python/paddle/reader/decorator.py (hunk 11 above)
Comment (helinwang) [inline python/paddle/reader/decorator.py]: Done.

File: python/paddle/reader/decorator.py (hunk 12 above)
Comment (helinwang) [inline python/paddle/reader/decorator.py]: Done.

**Comment by helinwang** [comment]
@reyoung Thanks, all changes done.

**Pull request merged**